llvm/lib/Target/AMDGPU/AMDGPUSwLowerLDS.cpp

```diff
@@ -0,0 +1,877 @@
+      "llvm.amdgcn.sw.lds.offset.table", nullptr, GlobalValue::NotThreadLocal,
+      AMDGPUAS::CONSTANT_ADDRESS);
+  return;
+}
+
+void AMDGPUSwLowerLDS::lowerNonKernelLDSAccesses(
+    Function *Func, SetVector<GlobalVariable *> &LDSGlobals,
+    NonKernelLDSParameters &NKLDSParams) {
+  // Replace LDS access in non-kernel with replacement queried from
+  // Base table and offset from offset table.
+  auto *EntryBlock = &Func->getEntryBlock();
+  IRB.SetInsertPoint(EntryBlock, EntryBlock->begin());
+  Function *Decl =
+      Intrinsic::getDeclaration(&M, Intrinsic::amdgcn_lds_kernel_id, {});
+  auto *KernelId = IRB.CreateCall(Decl, {});
+  GlobalVariable *LDSBaseTable = NKLDSParams.LDSBaseTable;
+  GlobalVariable *LDSOffsetTable = NKLDSParams.LDSOffsetTable;
+  auto &OrdereLDSGlobals = NKLDSParams.OrdereLDSGlobals;
+  assert(LDSBaseTable && LDSOffsetTable);
+  Value *BaseGEP = IRB.CreateInBoundsGEP(
+      LDSBaseTable->getValueType(), LDSBaseTable, {IRB.getInt32(0), KernelId});
+  Value *BaseLoad = IRB.CreateLoad(IRB.getInt32Ty(), BaseGEP);
+
+  for (GlobalVariable *GV : LDSGlobals) {
+    Value *BasePtr = IRB.CreateIntToPtr(BaseLoad, GV->getType());
+    auto GVIt = std::find(OrdereLDSGlobals.begin(), OrdereLDSGlobals.end(), GV);
+    assert(GVIt != OrdereLDSGlobals.end());
+    uint32_t GVOffset = std::distance(OrdereLDSGlobals.begin(), GVIt);
+    Value *OffsetGEP = IRB.CreateInBoundsGEP(
+        LDSOffsetTable->getValueType(), LDSOffsetTable,
+        {IRB.getInt32(0), KernelId, IRB.getInt32(GVOffset)});
+    Value *OffsetLoad = IRB.CreateLoad(IRB.getInt32Ty(), OffsetGEP);
+    OffsetLoad = IRB.CreateIntToPtr(OffsetLoad, GV->getType());
+    OffsetLoad = IRB.CreateLoad(IRB.getInt32Ty(), OffsetLoad);
+    Value *BasePlusOffset =
+        IRB.CreateInBoundsGEP(IRB.getInt8Ty(), BasePtr, {OffsetLoad});
+    replacesUsesOfGlobalInFunction(Func, GV, BasePlusOffset);
+  }
+  return;
+}
+
+static void reorderStaticDynamicIndirectLDSSet(KernelLDSParameters &LDSParams) {
+  // Sort Static, dynamic LDS globals which are either
+  // direct or indirect access on basis of name.
+  LDSParams.DirectAccess.StaticLDSGlobals =
+      sortByName(std::vector<GlobalVariable *>(
+          LDSParams.DirectAccess.StaticLDSGlobals.begin(),
+          LDSParams.DirectAccess.StaticLDSGlobals.end()));
+  LDSParams.DirectAccess.DynamicLDSGlobals =
+      sortByName(std::vector<GlobalVariable *>(
+          LDSParams.DirectAccess.DynamicLDSGlobals.begin(),
+          LDSParams.DirectAccess.DynamicLDSGlobals.end()));
+  LDSParams.IndirectAccess.StaticLDSGlobals =
+      sortByName(std::vector<GlobalVariable *>(
+          LDSParams.IndirectAccess.StaticLDSGlobals.begin(),
+          LDSParams.IndirectAccess.StaticLDSGlobals.end()));
+  LDSParams.IndirectAccess.DynamicLDSGlobals =
+      sortByName(std::vector<GlobalVariable *>(
+          LDSParams.IndirectAccess.DynamicLDSGlobals.begin(),
+          LDSParams.IndirectAccess.DynamicLDSGlobals.end()));
```

Comment:
ditto, can add
```
auto &DirectLDS = LDSParams.IndirectAccess;
auto &IndirectLDS = LDSParams.IndirectAccess;
```
to eliminate quite a bit of verbosity

I also wonder if you can eliminate the `std::vector<GlobalVariable *>(..)` and just use `{}` ? Sometimes it works, not sure if it'll work because of the `&&` arg though.